File: CPlusPlusTracer/CPlusPlusTracer/Color.cpp
#include "Color.h"

CColor::CColor(double r, double g, double b) : R(r), G(g), B(b) {
}

CColor CColor::Add(CColor c) {
   return CColor(this->R + c.R, this->G + c.G, this->B + c.B);
}

CColor CColor::Subtract(CColor c) {
   return CColor(this->R - c.R, this->G - c.G, this->B - c.B);
}

CColor CColor::Scale(double s) {
   return CColor(this->R * s, this->G * s, this->B * s);
}

CColor CColor::Multiply(CColor c) {
   return CColor(this->R * c.R, this->G * c.G, this->B * c.B);
}

int CColor::Legalize(double d) {
   return d > 1.0 ? 255 : (int)(d * 255.0);
}

int CColor::ToColor() {
   int r = this->Legalize(this->R);
   int g = this->Legalize(this->G);
   int b = this->Legalize(this->B);
   return (255 << 24) + (b << 16) + (g << 8) + r;
}

File: SwiftTracer/SwiftTracer/Color.swift
import Foundation
import Cocoa
import AppKit

struct Color {
    let r,g,b: Double;
    init(r:Double, g:Double, b:Double) {
        self.r = r;
        self.g = g;
        self.b = b;
    }
    
    func Add(c:Color) -> Color {
        return Color(r:self.r + c.r, g:self.g + c.g, b:self.b + c.b)
    }
    
    func Subtract(c:Color) -> Color {
        return Color(r:self.r - c.r, g:self.g - c.g, b:self.b - c.b)
    }
    
    func Scale(s:Double) -> Color {
        return Color(r:self.r * s, g:self.g * s, b:self.b * s)
    }
    
    func Multiply(c:Color) -> Color {
        return Color(r:self.r * c.r, g:self.g * c.g, b:self.b * c.b)
    }
    
    func ToColor() -> NSColor {
        let legalize =  { (var i : Double) -> CGFloat in
            if i > 1.0 { return CGFloat(1.0) }
            else { return CGFloat(i) }
        };
        
        let r = legalize(self.r)
        let g = legalize(self.g)
        let b = legalize(self.b)
        let color = NSColor(calibratedRed: r, green: g, blue: b, alpha: 1.0)

        return color;
    }
}